### PR TITLE
Support both Spray Json & Circe.  Add Msp Sessions module.

### PR DESCRIPTION
**Pull request by kiequoo**

### CODE REVIEW

File: endpoints-cpo-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsClient.scala [hunk 1]
```diff
@@ -4,25 +4,37 @@ import scala.concurrent.{ExecutionContext, Future}
...
+import com.thenewmotion.ocpi.msgs.{AuthToken, SuccessResp}
 import com.thenewmotion.ocpi.msgs.Ownership.Ours
 import com.thenewmotion.ocpi.msgs.v2_1.Locations._
 import cats.syntax.either._
-import com.thenewmotion.ocpi.msgs
-import spray.json.JsonFormat
 
-class MspLocationsClient(implicit http: HttpExt) extends OcpiClient {
+class MspLocationsClient(
+  implicit http: HttpExt,
```
Comment (DStranger) [inline endpoints-cpo-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsClient.scala]: Jesus Christ!
Can this be put into a trait or something?

File: endpoints-msp-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsRoute.scala [hunk 2]
```diff
@@ -1,24 +1,36 @@
...
 import msgs.v2_1.Locations._
 import msgs._
 import msgs.OcpiStatusCode._
+
 import scala.concurrent.ExecutionContext
 
 class MspLocationsRoute(
   service: MspLocationsService
+)(
+  implicit locationU: FromEntityUnmarshaller[Location],
```
Comment (DStranger) [inline endpoints-msp-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsRoute.scala]: Looks like implicit hell...

File: endpoints-msp-tokens/src/main/scala/com/thenewmotion/ocpi/tokens/MspTokensRoute.scala [hunk 3]
```diff
@@ -45,7 +48,11 @@ class MspTokensRoute(
 
   private val TokenUidSegment = Segment.map(TokenUid(_))
 
-  def route(apiUser: GlobalPartyId)(implicit ec: ExecutionContext) =
+  def route(
```
Comment (DStranger) [inline endpoints-msp-tokens/src/main/scala/com/thenewmotion/ocpi/tokens/MspTokensRoute.scala]: These format changes don't help understanding the commit :(

File: msgs/src/main/scala/com/thenewmotion/ocpi/msgs/v2_1/Sessions.scala [hunk 4]
```diff
@@ -34,9 +34,9 @@ object Sessions {
   }
 
   case class Session(
-    sessionId: SessionId,
-    startDateTime: ZonedDateTime,
-    endDateTime: ZonedDateTime,
+    id: SessionId,
+    startDatetime: ZonedDateTime,
```
Comment (DStranger) [inline msgs/src/main/scala/com/thenewmotion/ocpi/msgs/v2_1/Sessions.scala]: Hmm, don't like this serialisation-induced renaming, to be honest.

File: endpoints-cpo-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsClient.scala (hunk 1 above)
Comment (kiequoo) [reply endpoints-cpo-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsClient.scala]: Yeah, I don't like it either, but can't think of any other way to do this and doesn't tie it to a specific JSON lib.  Or we could just get rid of spray-json altogether ;)

File: msgs/src/main/scala/com/thenewmotion/ocpi/msgs/v2_1/Sessions.scala (hunk 4 above)
Comment (kiequoo) [reply msgs/src/main/scala/com/thenewmotion/ocpi/msgs/v2_1/Sessions.scala]: Better than having to declare all the field names IMHO...

File: endpoints-msp-sessions/src/main/scala/com/thenewmotion/ocpi/sessions/SessionsService.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * All methods are to be implemented in an idempotent fashion.
+  */
+trait SessionsService {
+
+  /**
+    * @return true if the session has been created and false if it has been updated
```
Comment (DStranger) [inline endpoints-msp-sessions/src/main/scala/com/thenewmotion/ocpi/sessions/SessionsService.scala]: A dedicated type would be better, I think.

Comment (kiequoo) [reply endpoints-msp-sessions/src/main/scala/com/thenewmotion/ocpi/sessions/SessionsService.scala]: Agreed, I based it it other modules.  I'll update it also in those then for consistency.

File: msgs-spray-json/src/main/scala/com/thenewmotion/ocpi/msgs/sprayjson/v2_1/LocationsJsonProtocol.scala [hunk 6]
```diff
@@ -1,13 +1,120 @@
...
+    suboperator: Option[BusinessDetails],
+    owner: Option[BusinessDetails],
+    facilities: Option[Iterable[Facility]],
+    timeZone: Option[String],
+    openingTimes: Option[Hours],
+    chargingWhenClosed: Option[Boolean],
+    images: Option[Iterable[Image]],
+    energyMix: Option[EnergyMix]
+  ) =
+    new Location(
```
Comment (DStranger) [inline msgs-spray-json/src/main/scala/com/thenewmotion/ocpi/msgs/sprayjson/v2_1/LocationsJsonProtocol.scala]: No need for a 'new' here

File: msgs-spray-json/src/main/scala/com/thenewmotion/ocpi/msgs/sprayjson/v2_1/TokensJsonProtocol.scala [hunk 7]
```diff
@@ -4,10 +4,22 @@ package sprayjson.v2_1
...
 import sprayjson.SimpleStringEnumSerializer
 import spray.json.{JsString, JsValue, JsonFormat, RootJsonFormat, deserializationError}
 
 trait TokensJsonProtocol {
+
+  private def deserializeLocationReferences(
+    locationId: LocationId,
+    evseUids: Option[Iterable[EvseUid]],
+    connectorIds: Option[Iterable[ConnectorId]]
+  ) = new LocationReferences(
```
Comment (DStranger) [inline msgs-spray-json/src/main/scala/com/thenewmotion/ocpi/msgs/sprayjson/v2_1/TokensJsonProtocol.scala]: 'new' is not required here

File: endpoints-cpo-locations/src/main/scala/com/thenewmotion/ocpi/locations/CpoLocationsRoute.scala [hunk 8]
```diff
@@ -37,34 +54,33 @@ class CpoLocationsRoute(
     em.compose[LocationsError] { locationsError =>
       val statusCode = locationsError match {
         case (_: LocationNotFound | _: EvseNotFound | _: ConnectorNotFound) => NotFound
-        case _ => InternalServerError
+        case _                                                              => InternalServerError
```
Comment (DStranger) [inline endpoints-cpo-locations/src/main/scala/com/thenewmotion/ocpi/locations/CpoLocationsRoute.scala]: This seems a bit unnecessary

File: endpoints-msp-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsRoute.scala [hunk 9]
```diff
@@ -29,22 +47,23 @@ class MspLocationsRoute(
...
     implicit em: ToResponseMarshaller[(StatusCode, ErrorResp)]
   ): ToResponseMarshaller[LocationsError] = {
-      em.compose[LocationsError] { locationsError =>
+    em.compose[LocationsError] { locationsError =>
       val statusCode = locationsError match {
-        case (_: LocationNotFound | _: EvseNotFound | _: ConnectorNotFound) => NotFound
+        case (_: LocationNotFound | _: EvseNotFound | _: ConnectorNotFound)                   => NotFound
         case (_: LocationCreationFailed | _: EvseCreationFailed | _: ConnectorCreationFailed) => OK
-        case _ => InternalServerError
+        case _                                                                                => InternalServerError
```
Comment (DStranger) [inline endpoints-msp-locations/src/main/scala/com/thenewmotion/ocpi/locations/MspLocationsRoute.scala]: This too =)

**Comment by DStranger** [comment]
I like the shapeless stuff you did, very cool 👍

**Pull request merged**